docs/topics/classes.md

```diff
@@ -1,245 +1,747 @@
+// Subclass inheriting from 'Person' and overriding the 'introduce()' function
+class Student(
+    name: String,
+    val school: String
+) : Person(name) {
+    override fun introduce() {
+        println("Hi, I'm $name, and I study at $school.")
+    }
 }
 ```
 
+In addition, you need to add the `open` keyword to every property
+or function that can be overridden.
+
+If you override a member of a base class, the overriding member
+will also be open by default. If you want to change this and forbid the subclasses of your
+class from overriding your implementation, you can explicitly mark the overriding
+member as `final`:
+
+```kotlin
+// Base class with the 'open' keyword to allow inheritance
+open class Person(val name: String) {
+    // Open function that can be overridden in a subclass
+    open fun introduce() {
+        println("Hello, my name is $name.")
+    }
+}
+
+// Subclass inheriting from 'Person' and overriding the 'introduce()' function
+class Student(name: String, val school: String) : Person(name) {
+    // The 'final' keyword prevents further overriding in subclasses
+    final override fun introduce() {
+        println("Hi, I'm $name, and I study at $school.")
+    }
+}
+```
+
+You don't need to annotate abstract classes or functions with the `open` keyword because they are implicitly 
+open and designed to be inherited. 
+
+Similarly, you can override a non-abstract `open` member with an abstract one:
+
+```kotlin
+// Abstract class with a primary constructor that defines 'name'
+abstract class Person(val name: String) {
+  // Non-abstract open member that allows overriding
+  open fun greet() {
+    println("Hello, my name is $name.")
+  }
+
+  // Abstract member
+  abstract fun introduce()
+}
+
+// Subclass that overrides the 'greet()' function with an abstract one
+abstract class Student(name: String, val school: String) : Person(name) {
+  // Overrides the non-abstract 'greet()' function with an abstract one
+  abstract override fun greet()
+
+  // Overrides the abstract 'introduce()' function with an abstract one
+  override fun introduce() {
+    println("I am $name, and I study at $school.")
+  }
+}
+
+```
+
 ## Companion objects
 
-If you need to write a function that can be called without having a class instance but that needs access to the internals
-of a class (such as a factory method), you can write it as a member of an [object declaration](object-declarations.md) inside that class.
+In Kotlin, a [companion object](object-declarations.md#companion-objects) is a type of object declaration 
+that allows you to access its members using the class name without needing an object from the class.
 
-Even more specifically, if you declare a [companion object](object-declarations.md#companion-objects) inside your class,
+If you need to write a function that can be called without having a class object but still needs access 
+to the internals of the class, you can define it inside `companion object` within that class:
```

Comment:
How do you define "internals of the class" here? A companion object can access its own internals, but class itself does not have any internals until you have created an instance of it.

Reply:
Also, here the terminology begins being a bit confusing, as "class object" and "companion object" look very similar, but mean significantly different things